### PR TITLE
[arm] Correctly detect hardware features when running as arm32 on arm64 hardware.

### PR DESCRIPTION
**Pull request by alexrp**

### CODE REVIEW

**Comment by alexrp** [comment]
@vargaz can you review?

File: mono/utils/mono-hwcap-arm.c [hunk 1]
```diff
@@ -150,6 +157,23 @@ mono_hwcap_arch_init (void)
...
+			mono_hwcap_arm_is_v6 = TRUE;
+			mono_hwcap_arm_is_v7 = TRUE;
+
+			mono_hwcap_arm_has_vfp = TRUE;
+			mono_hwcap_arm_has_vfp3 = TRUE;
+			mono_hwcap_arm_has_vfp3_d16 = TRUE;
+
+			mono_hwcap_arm_has_thumb = TRUE;
+			mono_hwcap_arm_has_thumb2 = TRUE;
+		}
```
Comment (vargaz) [inline mono/utils/mono-hwcap-arm.c]: Is this true in all cases ?

Comment (alexrp) [inline mono/utils/mono-hwcap-arm.c]: Judging by the instructions listed in the AArch32 section of the ARMv8 manual, I'd say assuming ARMv7 is safe. Judging by E1.2.1 and E1.3.2, I'd say assuming full VFPv3 support is also safe.

Thumb I'm not actually sure about. Any idea?

**Comment by lewurm** [comment]
isn't that something that should be reported by the kernel?
seems to be a bit inconsistent. on my android arm64 device:

```
u0_a112@clark:/ $ uname -a
Linux localhost 3.10.84-perf-g6898f68 #1 SMP PREEMPT Sat Jan 30 14:35:02 CST 2016 armv8lu0_a112@clark:/ $ cat /proc/cpuinfo
processor       : 0
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt evtstrm
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 3

[...]

Revision        : 83c0
Serial          : b55f92001e000000
Device          : clark
Radio           : 2
MSM Hardware    : MSM8992```

but then on one of our `aarch64` boards:
```[lewurm@apm_arm64 ~]$ cat /proc/cpuinfo
Processor	: AArch64 Processor rev 0 (aarch64)
processor	: 0
model name	: APM X-Gene Mustang board @ 2.40GHz
cpu MHz		: 2400.00
fpu		: yes

[...]

Features	: fp asimd
CPU implementer	: 0x50
CPU architecture: AArch64
CPU variant	: 0x0
CPU part	: 0x000
CPU revision	: 0

Hardware	: APM X-Gene Mustang board```

hm.  What's your intend for this change?

**Comment by alexrp** [comment]
The intention is to fix this bug: https://bugzilla.xamarin.com/show_bug.cgi?id=41596

This is not about detecting features of the 64-bit instruction set. It's about running 32-bit apps on a 64-bit machine and us failing to detect *any* 32-bit features whatsoever because we see a 64-bit `/proc/cpuinfo` which we currently don't parse as anything meaningful in the 32-bit feature detection code.

**Comment by lewurm** [comment]
@alexrp: I don't see how the current code would fail on my arm64 android device (Moto X Pure), because the kernel correctly reports the supported CPU features.

Anyway, apparently `/proc/cpuinfo` doesn't report consistently across devices, thus LGTM.

**Comment by alexrp** [comment]
@lewurm Could you provide the full output of `/proc/cpuinfo` on that Android device?

File: mono/utils/mono-hwcap-arm.c (hunk 1 above)
Comment (alexrp) [inline mono/utils/mono-hwcap-arm.c]: OK, digging through some ARM docs, it seems Thumb and Thumb-2 are both supported in ARMv8. See here: http://www.arm.com/products/processors/armv8-architecture.php

**Comment by alexrp** [comment]
I rewrote the patch to use `uname` to detect ARMv8 instead, as the `/proc/cpuinfo` format is wildly inconsistent across kernel versions, to the point of being useless.

**Comment by lewurm** [comment]
LGTM

**Comment by alexrp** [comment]
@vargaz does this look OK to you (in particular the assumption that all 32-bit features exist on 64-bit)?

**Comment by vargaz** [comment]
Looks ok.

**Comment by alexrp** [comment]
Merging manually since there are some unrelated test failures.

**Pull request merged**